apps/bfd-pipeline/bfd-pipeline-rda-grpc/src/main/java/gov/cms/bfd/pipeline/rda/grpc/AbstractCleanupJob.java

```diff
@@ -0,0 +1,162 @@
+   *
+   * @return the database table name for the parent entity.
+   */
+  abstract String getParentTableName();
+
+  /**
+   * Returns the key column of the database table mapped to the parent entity class.
+   *
+   * @return the key column name for the parent entity.
+   */
+  abstract String getParentTableKey();
+
+  /**
+   * Executes the job if enabled. Calculates the number of transactions from the cleanupRunSize /
+   * cleanupTransactionSize then for each transaction it determines the maximum last updated date
+   * for claims that will be removed then deletes them.
+   *
+   * <p>If any iteration fails to find claims to delete or the total number of claims deleted
+   * already exceeds the cleanupRunSize then processing stops.
+   *
+   * @return the number of deleted claims, or zero if not enabled.
+   */
+  public int run() throws ProcessingException {
+    int claimsDeleted = 0;
+
+    if (enabled) {
+      final long startMillis = System.currentTimeMillis();
+      var entityManager = entityManagerFactory.createEntityManager();
+
+      try {
+        Instant cutoffDate = Instant.now().minus(OLDEST_CLAIM_AGE_IN_DAYS, ChronoUnit.DAYS);
+        List<Query> queries = buildDeleteQueries(cutoffDate, entityManager);
+        var numberOfTransactions = Math.ceilDiv(cleanupRunSize, cleanupTransactionSize);
+
+        for (int i = 0; i < numberOfTransactions; i++) {
+          var result = executeDeleteTransaction(queries, entityManager);
+          claimsDeleted += result;
+
+          // If no claims deleted this iteration, or total claims deleted
+          // exceeds the cleanupRunSize, then stop
+
+          if (result == 0 || claimsDeleted >= cleanupRunSize) {
+            break;
+          }
+        }
+        final long endMillis = System.currentTimeMillis();
+        logger.info(
+            "cleanup job removed {} claims in {}ms", claimsDeleted, endMillis - startMillis);
+      } catch (Exception ex) {
+        logger.error("cleanup job aborted by an exception: message={}", ex.getMessage(), ex);
+        throw new ProcessingException(ex, 0);
+      } finally {
+        entityManager.close();
+      }
+    }
+
+    return claimsDeleted;
+  }
+
+  /**
+   * Executes a single delete transaction. The list of queries are passed in execution order, last
+   * query must delete from the parent table, all other queries are child table deletes.
+   *
+   * @param queries the queries to execute in each transaction.
+   * @param em the entitymanager to use with the deletion.
+   * @return the number of claims deleted by the transaction.
+   */
+  private int executeDeleteTransaction(List<Query> queries, EntityManager em) {
+    var count = 0;
+    em.getTransaction().begin();
+
+    // execute all queries in order, use the count from the last query (parent table)
+    // to track the number of claims deleted.
+    for (Query query : queries) {
+      count = query.executeUpdate();
+    }
+    em.getTransaction().commit();
```

Comment:
This is exactly what `TransactionManager` was designed to do.

Reply:
Done.